### PR TITLE
Purge old realmless users

### PR DESCRIPTION
**Pull request by whaught**
Fixes https://github.com/google/exposure-notifications-verification-server/issues/606

## Proposed Changes

* Delete users who have been removed from all realms, are not a system admin, and are older than USER_PURGE_MAX_AGE

**Release Note**

```release-note
Cleanup for old users who have no realms and have aged out. This clears their DB information, but not their auth.
```

### CODE REVIEW

**Comment by whaught** [comment]
I tried to do this as

```
Joins("LEFT JOIN user_realms ON users.id = user_realms.user_id").
Where("AND user_realms.user_id IS NULL").
```

but that didn't seem to work

File: pkg/database/user.go [hunk 1]
```diff
@@ -330,6 +330,21 @@ func (db *Database) DeleteUser(u *User, actor Auditable) error {
...
+// PurgeUsers will delete users who are not a system admin, not a member of any realms
+// and have not been modified before the expiry time.
+func (db *Database) PurgeUsers(maxAge time.Duration) (int64, error) {
+	if maxAge > 0 {
+		maxAge = -1 * maxAge
+	}
+	deleteBefore := time.Now().UTC().Add(maxAge)
+	// Delete users who were created/updated before the expiry time.
+	rtn := db.db.Unscoped().
+		Where("users.system_admin = false AND users.created_at < ? AND users.updated_at < ?", deleteBefore, deleteBefore).
```
Comment (sethvargo) [inline pkg/database/user.go:342]: I don't think this does what you think it does? 

- Realm membership changes don't change `updated_at`, so you could end up immediately deleting someone
- That column is updated anytime the direct user record changes, we can happen at login with the revoke check

**Review by sethvargo — COMMENTED** [review]
/hold

File: pkg/database/user.go (hunk 1 above)
Comment (whaught) [reply pkg/database/user.go:342]: The latter isn't a problem - it's okay to not-delete users who are logging in to keep fresh.

I suppose we can ensure users is bumped on realm-remove

Comment (sethvargo) [reply pkg/database/user.go:342]: > I suppose we can ensure users is bumped on realm-remove

I couldn't figure out a good way to do that. If we can "touch" the update_at on the user when realm membership is changed, I think this is fine (maybe add a comment).

Comment (whaught) [reply pkg/database/user.go:342]: I've added `TestRemoveRealmUpdatesTime` and I can't seem to repro your claim that modifying the realm relationship _isn't_ bumping time

File: pkg/database/user_test.go [hunk 2]
```diff
@@ -97,6 +97,115 @@ func TestUserLifecycle(t *testing.T) {
...
+		Name:        "Dr Delete",
+		SystemAdmin: true,
+	}
+
+	if err := db.SaveUser(&user, System); err != nil {
+		t.Fatalf("error creating user: %v", err)
+	}
+	expectExists(t, db, user.ID)
+
+	db.PurgeUsers(time.Duration(0)) // is admin
```
Comment (sethvargo) [inline pkg/database/user_test.go:117]: Check error

File: pkg/database/user.go (hunk 1 above)
Comment (sethvargo) [reply pkg/database/user.go:342]: Interesting... what happens if the user comes serialized from the cache?

**Comment by mikehelmick** [comment]
not sure if we can use it - but firebase stores the last login date of a user.

File: pkg/database/user.go (hunk 1 above)
Comment (whaught) [reply pkg/database/user.go:342]: I believe the reason is that we have wrapped calls with SaveUser here:
https://github.com/google/exposure-notifications-verification-server/blob/f76a3cea09d53d076a47a204ce9c948e183c8f35/pkg/database/user.go#L337

which always saves both the User entry and the associations (even if the table fields haven't changed)

**Review by sethvargo — APPROVED** [review]
/lgtm
/hold 

Can you update the release note to warn about this and note that it does _not_ delete them from the upstream auth provider?

**Comment by sethvargo** [comment]
Hmm - you might need to rebase @whaught

**Pull request merged**